src/wasm/wasm-type.cpp

```diff
@@ -3306,13 +3653,39 @@ std::optional<TypeBuilder::Error> canonicalizeIsorecursive(
     return {*error};
   }
 
-  // Move the recursion groups into the global store. TODO: after proper
-  // isorecursive canonicalization, some groups may no longer be used, so they
-  // will need to be filtered out.
-  std::lock_guard<std::mutex> lock(recGroupsMutex);
-  for (auto& info : recGroupInfos) {
-    recGroups.emplace_back(std::move(info));
+  // Now that we know everything is valid, start canonicalizing recursion
+  // groups. Before canonicalizing each group, update all the use sites with it
+  // to make sure it only refers to other canonical groups or to itself. To
```

Comment:
How could the uses point to anything else? We have already validated that they do not refer to any later rec group, I think - or am I wrong?

Reply:
If recursion group `B` refers to recursion group `A`, then yes, we must have already canonicalized `A`. But perhaps `A` was equivalent to some previously canonicalized recursion group `A'`. In that case, we would have thrown out `A` and we need to replace all references to `A` with equivalent references to `A'`.

Reply:
Oh, wow, I didn't even think about that. Yes, makes sense.

Maybe worth clarifying in the comment? Might be just me though...

Reply:
Sure, I'll call this out more specifically in the comment.

Reply:
done.